src/tools/rustfmt/src/patterns.rs

```diff
@@ -135,7 +135,7 @@ impl Rewrite for Pat {
                 let mut_prefix = format_mutability(mutability).trim();
 
                 let (ref_kw, mut_infix) = match by_ref {
-                    ByRef::Yes(rmutbl) => ("ref", format_mutability(rmutbl).trim()),
+                    ByRef::Yes(_, rmutbl) => ("ref", format_mutability(rmutbl).trim()),
```

Comment:
Admittedly I've not done a full review of the entire diff, but I believe this would result in the `pin` being dropped here in the `Pinnedness::Pinned` case wouldn't it?

Reply:
Sure, but parsing has not been implemented yet, and thus, `ref pin` cannot appear in the code yet.

BTW, this PR is currently WIP, and I'd just use this temp commit to discuss some questions with some earlier reviewers, because I'm not quite sure about the design of the pattern analysis part.

Reply:
I added a `// FIXME` comment here.